internal/controllers/token_exchange.go

```diff
@@ -285,13 +259,56 @@ func (t *TokenExchangeController) evaluateSacdDoc(c *fiber.Ctx, record *models.P
 
 	// If any permissions are missing, return an error
 	if len(missingPermissions) > 0 {
-		return fiber.NewError(fiber.StatusBadRequest,
-			fmt.Sprintf("Address %s lacks permissions %v on token id %d for asset %s.",
-				grantee.Hex(), missingPermissions, pr.TokenID, pr.NFTContractAddress))
+		return fmt.Errorf("missing permissions: %v on token id %d for asset %s", missingPermissions, pr.TokenID, pr.NFTContractAddress)
 	}
 
-	// If we get here, all permissions are valid
-	return t.createAndReturnToken(c, pr, grantee)
+	return nil
+}
+
+func (t *TokenExchangeController) evaluateAttestations(record *models.PermissionRecord, tokenReq *PermissionTokenRequest) error {
+	if valid, err := t.validateAssetDID(record.Data.Asset, tokenReq); err != nil || !valid {
+		return fmt.Errorf("failed to validate attestation asset: %s", record.Data.Asset)
+	}
+
+	attestations := make(map[string]map[string]struct{})
+	for _, agreement := range record.Data.Agreements {
+		if agreement.EventType != "dimo.attestation" {
+			continue
+		}
+
+		if agreement.ExpiresAt.Before(time.Now()) || time.Now().Before(agreement.EffectiveAt) {
+			continue
+		}
+
+		attestations[agreement.Source] = make(map[string]struct{})
+		for _, id := range agreement.ID {
+			attestations[agreement.Source][id] = struct{}{}
+		}
+	}
+
+	var errs []error
+	for _, claim := range tokenReq.Attestations {
+		att, ok := attestations[claim.Source]
+		if !ok {
+			errs = append(errs, fmt.Errorf("missing grant for source: %s", claim.Source))
+			continue
+		}
+
+		var missing []string
+		for _, attID := range claim.AttestationIDs {
+			_, ok := att[attID]
+
+			if !ok {
+				missing = append(missing, attID)
+			}
+		}
+
+		if len(missing) >= 1 {
+			errs = append(errs, fmt.Errorf("for source %s missing grants for attestation ids: %s", claim.Source, strings.Join(missing, ", ")))
+		}
+	}
+
+	return errors.Join(errs...)
```

Comment:
You don't need a list for this
```suggestion
	var errs error
	for _, claim := range tokenReq.Attestations {
		att, ok := attestations[claim.Source]
		if !ok {
			errs = append(errs, fmt.Errorf("missing grant for source: %s", claim.Source))
			continue
		}

		var missing []string
		for _, attID := range claim.AttestationIDs {
			_, ok := att[attID]

			if !ok {
				missing = append(missing, attID)
			}
		}

		if len(missing) >= 1 {
			err = errors.Join(fmt.Errorf("for source %s missing grants for attestation ids: %s", claim.Source, strings.Join(missing, ", "))
		}
	}

	return errs
```